### PR TITLE
SQL Download Feature

### PR DESCRIPTION
**Pull request by aalbatross**
For https://github.com/gbif/occurrence/issues/71

1. Support for SQL Download
2. SQL query validation
3. adding endpoints, to describe, validate and accept SQL download request.

### CODE REVIEW

File: occurrence-download/src/main/resources/templates/simple-sql-download/execute-simple-sql-query.ftl [hunk 1]
```diff
@@ -0,0 +1,39 @@
+<#--
```
Comment (fmendezh) [inline occurrence-download/src/main/resources/templates/simple-sql-download/execute-simple-sql-query.ftl]: this file does not need to be a template, can be moved to the hive-scripts dir, remove all ${r" and "}

File: occurrence-ws/src/main/java/org/gbif/occurrence/ws/provider/hive/ConnectionPool.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+import org.apache.nifi.reporting.InitializationException;
+import org.gbif.utils.file.properties.PropertiesUtil;
+import org.gbif.ws.app.ConfUtils;
+
+/**
+ * 
+ * ConnectionPool for Hive JDBC connection.
+ *
+ */
+public class ConnectionPool{
```
Comment (fmendezh) [inline occurrence-ws/src/main/java/org/gbif/occurrence/ws/provider/hive/ConnectionPool.java]: This class is being used incorrectly, a connection pool instance is created each time a validation is done, and instance should be  kept at the ws level or at the guice model

File: occurrence-ws/src/main/java/org/gbif/occurrence/ws/provider/hive/HiveSQL.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      public String transsql() {
+        return transsql;
+      }
+      
+      public String sqlHeader() {
+        return sqlHeader;
+      }
+
+      @Override
+      public String toString() {
```
Comment (fmendezh) [inline occurrence-ws/src/main/java/org/gbif/occurrence/ws/provider/hive/HiveSQL.java]: we should not rely on Json serialization in a toString method

File: occurrence-ws/src/main/java/org/gbif/occurrence/ws/provider/hive/Result.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        return comment;
+      }
+
+      public void setComment(String comment) {
+        this.comment = comment;
+      }
+
+      @Override
+      public String toString() {
+        StringBuilder builder = new StringBuilder();
```
Comment (fmendezh) [inline occurrence-ws/src/main/java/org/gbif/occurrence/ws/provider/hive/Result.java]: rely on json serialization in toString method

File: occurrence-ws/src/main/java/org/gbif/occurrence/ws/provider/hive/query/validator/QueryContext.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import org.apache.calcite.sql.SqlIdentifier;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.parser.SqlParser;
+import org.apache.calcite.sql.parser.SqlParser.Config;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.gbif.common.shaded.com.google.common.base.Preconditions;
+import org.gbif.occurrence.ws.provider.hive.query.validator.Query.Issue;
+
```
Comment (fmendezh) [inline occurrence-ws/src/main/java/org/gbif/occurrence/ws/provider/hive/query/validator/QueryContext.java]: javadoc

File: occurrence-ws/src/main/java/org/gbif/occurrence/ws/resources/DownloadResource.java [hunk 6]
```diff
@@ -106,9 +109,47 @@ public Response oozieCallback(@QueryParam("job_id") String jobId, @QueryParam("s
...
+  @POST
+  @Validate
+  @Path("sql")
+  public String startDownload(@Valid SqlDownloadRequest request, @Context SecurityContext security) {
+    LOG.debug("Download: [{}]", request);
+
+    // assert authenticated user is the same as in download
+    assertLoginMatches(request, security);
+
+    HiveSQL.Validate.Result result = new HiveSQL.Validate().apply(request.getSql());
```
Comment (fmendezh) [inline occurrence-ws/src/main/java/org/gbif/occurrence/ws/resources/DownloadResource.java]: maybe if we move this validation to the DownloadReqestServiceImpl we can avoid the artificial field sqlHeader

Comment (aalbatross) [reply occurrence-ws/src/main/java/org/gbif/occurrence/ws/resources/DownloadResource.java]: done.

**Comment by aalbatross** [comment]
Merging to master.

**Pull request merged**